cheatsheets/WebSocket_Security_Cheat_Sheet.md

```diff
@@ -0,0 +1,251 @@
+
+```javascript
+// Example: Close WebSocket on session expiry
+function validateSession(ws, sessionId) {
+  if (!isSessionValid(sessionId)) {
+    ws.close(1008, 'Session expired');
+    return false;
+  }
+  return true;
+}
+```
+
+**When users log out**, close all their WebSocket connections immediately. Maintain a mapping of sessions to active connections so you can invalidate WebSocket access the moment logout occurs.
+
+**Token refresh:**
+
+Rotate tokens in long-lived connections to prevent hijacked sessions from persisting.
+
+#### Message-Level Authorization
+
+Don't assume WebSocket connection equals unlimited access. Check authorization for each action:
+
+```javascript
+ws.on('message', (data) => {
+  const message = JSON.parse(data);
+  
+  // Check authorization for each action
+  if (message.action === 'delete_user' && !user.hasRole('admin')) {
+    ws.send(JSON.stringify({type: 'error', message: 'Access denied'}));
+    return;
+  }
+  
+  handleAuthorizedMessage(ws, user, message);
+});
+```
+
+### Input Validation
+
+Treat all WebSocket messages as untrusted input. WebSocket messages can carry injection payloads such as SQLi, XSS, and command injection.
+
+**Validate message structure and content** using JSON schemas and allow-lists. Set reasonable size limits (typically 64KB or less) and implement rate limiting to prevent message flooding.
+
+**For binary data**, verify file types using magic numbers rather than trusting content-type headers. Scan uploads for malware when appropriate, and use safe deserialization for protocols like protobuf or MessagePack.
+
+```javascript
+ws.on('message', (data, isBinary) => {
+  if (isBinary) {
+    // Validate binary data
+    if (data.length > MAX_BINARY_SIZE) {
+      ws.close(1009, 'Message too large');
+      return;
+    }
+    
+    // Check file type by magic numbers
+    if (!isValidFileType(data)) {
+      ws.close(1008, 'Invalid file type');
+      return;
+    }
+  }
+  
+  processBinaryData(data);
+});
+```
+
+**Always use `JSON.parse()` instead of `eval()`** for JSON processing - `eval()` enables code execution from untrusted input.
+
+```javascript
+// Safe
+const message = JSON.parse(data);
+
+// Dangerous - enables code execution
+// const message = eval('(' + data + ')');
+```
+
+See the [Input Validation Cheat Sheet](Input_Validation_Cheat_Sheet.md) for more details.
+
+### Service Tunneling Risks
+
+Don't tunnel TCP services (VNC, FTP, SSH) through WebSockets. If your application has XSS vulnerabilities, attackers could access these services directly from victims' browsers.
+
+### Denial-of-Service Protection
+
+Persistent WebSocket connections increase DoS risk.
+
+**Limit connections and resources** by restricting total connections and implementing per-IP limits. Set **message size limits** (typically 64KB or less) and implement **rate limiting** to prevent message flooding - 100 messages per minute is a common starting point.
```

Comment:
"per-IP" is possibly dubious (especially with IPv6…). In many cases, wouldn't it be better to use per-user limits instead?